File: sentil-cpp/tests/test_monitor.cpp
#include <sentil/sentil.hpp>

#include <vector>

#include "sentil_test.hpp"

using sentil::Config;
using sentil::Formula;
using sentil::FormulaBank;
using sentil::Monitor;
using sentil::MultiMonitor;
using sentil::OnlineMonitor;
using sentil::TimeMode;
using sentil::Trace;

int main() {
    Monitor m("always[0,2](x > 0)");
    Trace t = Trace::indexed(3);
    t.add_signal("x", {3, 1, 2});
    CHECK_BITS(m.robustness(t), 1.0);
    CHECK(m.robustness_signal(t) == std::vector<double>({1, 1, 2}));
    CHECK(m.config().time() == TimeMode::Discrete);
    CHECK(m.symbol_index("x").value() == 0);
    CHECK(!m.symbol_index("absent").has_value());

    Monitor fold("(x > 0) and (y > 0)");
    CHECK(fold.update(0, {{"x", 4.0}, {"y", 2.0}}).value == 2.0);
    std::vector<double> packed(2);
    packed[*fold.symbol_index("x")] = 4.0;
    packed[*fold.symbol_index("y")] = -1.0;
    CHECK(fold.update_packed(1, packed).value == -1.0);

    Monitor dense(Formula::parse("always(x > 0)"), Config(TimeMode::Dense));
    CHECK(dense.config().time() == TimeMode::Dense);

    OnlineMonitor online("x > 0");
    CHECK(online.variable_count() == 1);
    auto run = online.run(t);
    CHECK(run.size() == 3 && run[0].value == 3 && run[1].value == 1);

    MultiMonitor mm;
    mm.add("pos", "x > 0");
    mm.add("big", Formula::parse("x > 10"));
    CHECK(mm.ids() == std::vector<std::string>({"pos", "big"}));
    auto verdicts = mm.update(0, {{"x", 5.0}});
    CHECK(verdicts.at("pos").satisfied && !verdicts.at("big").satisfied);
    CHECK(mm.remove("big") && mm.size() == 1);

    FormulaBank bank;
    bank.add("a", "x > 0");
    bank.add("b", "always[0,1](x > 0)");
    auto results = bank.robustness(t);
    CHECK_BITS(results.at("a"), 3.0);
    CHECK_BITS(results.at("b"), 1.0);

    return sentil_report("test_monitor");
}

File: sentil-jl/test/test_monitor.jl
using Test
using Sentil

@testset "offline monitor" begin
    trace = Trace([0.0, 1.0, 2.0], "x", [3.0, 1.0, 2.0])
    m = Monitor("always[0, 2](x > 0)")
    @test robustness(m, trace) == 1.0
    @test robustness_signal(m, trace) == robustness_signal(formula("always[0,2](x>0)"), trace)
    @test time_mode(config(m)) == TimeMode.Discrete
    @test time_mode(config(Monitor("x > 0"; config = Config(time = TimeMode.Dense)))) == TimeMode.Dense
end

@testset "monitor update" begin
    m = Monitor("x > 0")
    r = update!(m, 0.0, Dict("x" => 2.5))
    @test r.value == 2.5 && r.satisfied
    @test symbol_index(m, "x") == 1
    @test symbol_index(m, "absent") === nothing
    reset!(m)
    @test update_packed!(m, 0.0, [4.0]).value == 4.0
end

@testset "online monitor" begin
    m = OnlineMonitor("x > 0")
    @test variable_count(m) == 1
    trace = Trace([0.0, 1.0, 2.0, 3.0], "x", [3.0, 1.0, 2.0, 4.0])
    verdicts = run!(m, trace)
    @test [v.value for v in verdicts] == [3.0, 1.0, 2.0, 4.0]
end

@testset "multi monitor" begin
    mm = MultiMonitor()
    add!(mm, "pos", "x > 0")
    add!(mm, "big", formula("x > 5"))
    @test ids(mm) == ["pos", "big"]
    v = update!(mm, 0.0, Dict("x" => 3.0))
    @test v["pos"].satisfied && !v["big"].satisfied
    @test remove!(mm, "big") && length(mm) == 1
end

@testset "formula bank" begin
    b = FormulaBank()
    add!(b, "a", "x > 0")
    add!(b, "b", "always[0,2](x > 1)")
    trace = Trace([0.0, 1.0, 2.0], "x", [3.0, 1.0, 2.0])
    r = robustness(b, trace)
    @test r["a"] == 3.0 && r["b"] == 0.0
    bad = FormulaBank()
    add!(bad, "u", "y > 0")
    @test_throws SentilError robustness(bad, trace)
end
